vexination/walker/src/server.rs

```diff
@@ -40,87 +49,138 @@ pub async fn run(
         client = client.add_root_certificate(reqwest::tls::Certificate::from_pem(&pem)?);
     }
 
-    let client = Arc::new(client.build()?);
-
-    let validation = ValidationVisitor::new(|advisory: Result<ValidatedAdvisory, ValidationError>| {
-        let sink = sink.clone();
-        let provider = provider.clone();
-        let client = client.clone();
-        async move {
-            match advisory {
-                Ok(ValidatedAdvisory {
-                    retrieved:
-                        RetrievedAdvisory {
-                            data,
-                            discovered: DiscoveredAdvisory { url, .. },
-                            ..
-                        },
-                }) => {
-                    let name = url.path_segments().and_then(|s| s.last()).unwrap_or_else(|| url.path());
-                    match serde_json::from_slice::<csaf::Csaf>(&data) {
-                        Ok(doc) => match client
+    let total = Arc::new(AtomicUsize::default());
+    let duplicates: Arc<std::sync::Mutex<Duplicates>> = Default::default();
+    let errors: Arc<std::sync::Mutex<BTreeMap<DocumentKey, String>>> = Default::default();
+    let warnings: Arc<std::sync::Mutex<BTreeMap<DocumentKey, Vec<CheckError>>>> = Default::default();
+
+    {
+        let client = Arc::new(client.build()?);
+        let total = total.clone();
+        let duplicates = duplicates.clone();
+        let errors = errors.clone();
+        let warnings = warnings.clone();
+
+        let visitor = move |advisory: Result<
+            VerifiedAdvisory<ValidatedAdvisory, &'static str>,
+            VerificationError<ValidationError, ValidatedAdvisory>,
+        >| {
+            (*total).fetch_add(1, Ordering::Release);
+
+            let errors = errors.clone();
+            let warnings = warnings.clone();
+            let client = client.clone();
+            let sink = sink.clone();
+            let provider = provider.clone();
+            async move {
+                let adv = match advisory {
+                    Ok(adv) => {
+                        let sink = sink.clone();
+                        let name = adv
+                            .url
+                            .path_segments()
+                            .and_then(|s| s.last())
+                            .unwrap_or_else(|| adv.url.path());
+                        match client
                             .post(sink)
                             .header(header::CONTENT_TYPE, "application/json")
-                            .body(data.clone())
+                            .body(serde_json::to_string(&adv.csaf.clone()).unwrap())
```

Comment:
Please don't `unwrap`, as this might panic!

Reply:
if it's panic, do we record as well and proceed?